doc/specification/language-proposal.md

```diff
@@ -0,0 +1,1518 @@
+
+### Value bindings
+<a id="value-bindings"/>
+
+> _value-binding_ ::= _let-binding_ | _with-binding_
+>
+> _let-binding_ ::= `let` ***symbol*** ( _type-assertion_? `=` _expression_ | `=` _function-literal_ ) `;` _expression_
+>
+> _with-binding_ ::= `with` _expression_ `;` _expression_
+
+Both `let` and `with` introduce new identifiers in expression context that can mask bindings from outer scopes.
+
+The value of the let expression `let id = expr₁; expr₂` is `expr₂` with `id` bound to `expr₁`.
+
+The value of the where expression `where expr₁; expr₂` is `expr₂` with the bindings `fᵢ = vᵢ` for each field `fᵢ` with value `vᵢ` in the record value of `expr₁`. If `expr₁` does not have record type, the expression is ill-formed.
+
+Example:
+
+```
+let a = { scale = 3.2; pos = { x = 3 m ; y = 4 m; }; };   # binds `a` below.
+with a.pos;   # binds `x` to 3 m and `y` to 4 m below.
+a.scale*(x+y);
+```
+
+As the type of a value binding is the type of its final expression, syntactic ambiguity in type assertion parsing has no semantic consequence: `let id = expr₁; expr₂: T` has the type of `expr₂: T`, which is `T` iff the type of `expr₂` is `T` or a supertype of `T` ; similarly, `(let id = expr₁; expr₂): T` has type `T` iff the type of `let id = expr₁; expr₂`, which is the type of `expr₂` is `T` or a supertype of `T`.
+
+
+### Conditional expressions
+<a id="conditional-expressions"/>
+
+> _conditional-expr_ ::= _if-expr_ | _case-expr_
+>
+> _if-expr_ ::= `if` _boolean-expr_ `then` _expression_ `else` _expression_
+>
+> _case-expr_ ::= `|` _boolean-expr_ ***right-arrow*** _expression_ )? _case-expr_ | `|` ( `otherwise` | `true` ) ***right-arrow*** _expression_
+
+Alternatives based on one or more conditions can be expressed with if/then/else clauses or case expressions introduced with the bar symbol. In either, the conditional expression is ill-formed if any of the _boolean-expr_ have type which is not boolean.
+
+The value of `if condition then expr₁ else expr₂` is the value of `expr₁` in the case where `condition` is true, and `expr₂` otherwise. The type of the _if-expr_ is the type of `expr₁` and `expr₂`; the expression is ill-formed if the types of `expr₁` and `expr₂` differ.
+
+The value of `| otherwise → expr` is just the value of `expr`. `true` can be used in place of `otherwise` equivalently. For a compound _case-expr_ of the form `| condition → expr₁ case₂`, the value is `expr₁` in the case where `condition` is true, and the value of `case₂` otherwise; the expression is ill-formed if the types of `expr₁` and `case₂` differ.
+
+Example:
+
+```
+# Equivalent to 10 km:
+let a = if 3>2 then 10 m else 2 m; a*1000
+
+# Piece-wise constant reaction rate function.
+def rate = fn (T: temperature) →
+  | T <  0 °C → 10 mM/s
+  | T < 10 °C → 20 mM/s
+  | otherwise → 30 mM/s;
+
+```
```

Comment:
Note on semantics:
1. We evaluate these top-down, so 
```
def rate = fn (T: temperature) →
  | T <    0 °C → 10 mM/s
  | T < -10 °C → 20 mM/s
  | otherwise → 30 mM/s;
 ```
 will never trigger the 2nd case.
 2. I would love these to be checked for exhaustiveness, but that might hard to implement

Reply:
Re: 2. The `otherwise/true` clause is part of the syntax, rather than e.g. a final clause
> `|` _expression_ → _expression_
where the first _expression_ must be true (or exhaustive). It just makes things easy for us and clear for them IMO.